appscale/tools/agents/azure_agent.py

```diff
@@ -143,6 +147,19 @@ class AzureAgent(BaseAgent):
   # The Storage Azure Resource provider namespace.
   MICROSOFT_STORAGE_RESOURCE = 'Microsoft.Storage'
 
+  # Resource Identifier Cache for Azure. Each Resource Identifier matches a
+  # Public IP Address, a Network Interface and a Virtual Machine.
+  RI_CACHE = Queue()
```

Comment:
Using a queue is only thread-safe in the sense that it makes common queue operations atomic (like checking if the queue is full before adding to it). Adding an item to and retrieving keys from a set, for example, is atomic in CPython, and you could always add a threading lock to be safe for different runtimes. The use of a queue in describe_instances seems a bit awkward to me since it needs to be emptied and repopulated just to see what's in the cache.